lib/pbench/server/database/models/datasets.py

```diff
@@ -1151,9 +1128,50 @@ def setvalue(dataset: Dataset, key: str, value: Any, user: Optional[User] = None
                     dataset, key, value, f"date/time before {maximum:%Y-%m-%d}"
                 )
             target += __class__.ONE_DAY
-            v = f"{target:%Y-%m-%d}"
+            return f"{target:%Y-%m-%d}"
         else:
-            v = value
+            return value
+
+    @staticmethod
+    def setvalue(dataset: Dataset, key: str, value: Any, user: Optional[User] = None):
+        """
+        Create or modify an existing metadata value. This method supports
+        hierarchical dotted paths like "dashboard.seen" and should be used in
+        most contexts where client-visible metadata paths are used.
+
+        This will create a nested JSON structure (represented as a Python dict)
+        as necessary as it descends the hierarchy. For example, if you assign
+        "a.b.c" a value of "bar", and then query "a", you'll get back
+        "a": {"b": {"c": "bar"}}}
+
+        We handle two keys specially:
+
+            dataset.name: The primary resource name of a dataset, which can be
+                modified by the dataset's owner.
+            server.deletion: The expected automatic deletion date of the
+                dataset. It can be modified by the owner so long as the new
+                value is a valid date that falls within the maximum server
+                retention period from the dataset's upload timestamp.
```

Comment:
`server.deletion` is no longer handled specially by this function; that has been delegated to `validate()`, so we probably shouldn't mention it here.

And, given that we have only one special case (which already has a nice block comment in the code), I would be inclined to remove this whole section of the docstring.